### ISSUE TITLE
sky model failures on 20210529

### ISSUE THREAD

**Issue by sbailey** [desihub/desispec#1864]
In the Himalayas run we had multiple failures of the new PCA sky model code for exposures that worked with Fuji.  An example is

```
desi_compute_sky -i /global/cfs/cdirs/desi/spectro/redux/himalayas/exposures/20210529/00090224/frame-r3-00090224.fits.gz --fiberflat /global/cfs/cdirs/desi/spectro/redux/himalayas/exposures/20210529/00090224/fiberflatexp-r3-00090224.fits.gz -o /global/cfs/cdirs/desi/users/$USER/temp/sky-r3-00090224.fits.gz --adjust-wavelength --adjust-lsf --pca-corr /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm6/skycorr-pca-sm6-r3.fits --fit-offsets --skygradpca /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm6/skygradpca-sm6-r3.fits --tpcorrparam /global/cfs/cdirs/desi/spectro/desi_spectro_calib/trunk/spec/sm6/tpcorrparam-sm6-r3.fits
...
Intel MKL ERROR: Parameter 4 was incorrect on entry to DGELSD.
Traceback (most recent call last):
  File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/code/desispec/main/bin/desi_compute_sky", line 17, in <module>
    sys.exit(sky.main(args))
  File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/code/desispec/main/py/desispec/scripts/sky.py", line 109, in main
    skymodel = compute_sky(frame,add_variance=(not args.no_extra_variance),\
  File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/code/desispec/main/py/desispec/sky.py", line 564, in compute_sky
    res = compute_sky_linear(
  File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/code/desispec/main/py/desispec/sky.py", line 373, in compute_sky_linear
    skytpcorrcoeff = np.linalg.lstsq(
  File "<__array_function__ internals>", line 5, in lstsq
  File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/conda/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 2306, in lstsq
    x, resids, rank, s = gufunc(a, b, rcond, signature=signature, extobj=extobj)
  File "/global/common/software/desi/perlmutter/desiconda/20220119-2.0.1/conda/lib/python3.9/site-packages/numpy/linalg/linalg.py", line 100, in _raise_linalgerror_lstsq
    raise LinAlgError("SVD did not converge in Linear Least Squares")
numpy.linalg.LinAlgError: SVD did not converge in Linear Least Squares
```

@schlafly could you take a look at this?

Failures also occured on exposures 90225, 90239, 90244, and suspiciously it is the same 24 cameras that fail on each of them so it might be something else like the nightly flat badly messed up.

**Comment by schlafly** [comment]
Inspecting now, thanks.

**Comment by schlafly** [comment]
This particular case gets cleaned up by replacing the means with nanmeans here:
https://github.com/desihub/desispec/blob/main/py/desispec/skygradpca.py#L56-L57
Presumably the 4 problematic FIBERASSIGN_X/Y are not sky fibers, and so the rest of the code doesn't see them unless all of the positions are NaNed out by that mean.  But I probably need to harden the code more generally against NaNs in FIBERASSIGN_X/Y.  Ideally I'd replace those with the nominal centers of the patrol radii---I worry that replacing them with 0 could lead to crazy TPCORR leading to problems downstream.  I'm not sure what the most convenient way to get access to those nominal centers is, though.

Presently the new sky code uses both FIBER_X/Y and FIBERASSIGN_X/Y; it's not clear that it needs to do that.  In this case, the FIBER_X/Y are 0 and the FIBERASSIGN_X/Y are NaN; for this code's purposes, both should be replaced with the nominal patrol radius.

**Comment by araichoor** [comment]
if relevant: for that case, those are un-assigned fibers, i.e. `(FIBERSTATUS & 1) > 0`.
all are stuck, ~half are broken.
https://github.com/desihub/fiberassign/blob/bf17ccb3ea8fb735579b1d0c9b37997fb8518c8f/src/hardware.h#L24-L28

maybe working only with assigned fibers, i.e. `(FIBERSTATUS & 1) == 0`, could be a solution?

**Comment by schlafly** [comment]
Yes, even just doing nanmean(...) successfully addresses this particular case.  But I'm having flashbacks to weird test tiles with lots of NaNs (maybe due to proper motion corrections, or something?), and I think I should probably handle it more completely.

**Comment by araichoor** [comment]
sure, if using nanmean() works, it s maybe fine.
as far as I recall, I think there were NaNs in PMRA,PMDEC for some early tiles (which PlateMaker didn t like...); but it could be that also happened for TARGET_RA,TARGET_DEC.

**Comment by schlafly** [comment]
I've opened a reasonable looking PR here:
https://github.com/desihub/desispec/pull/1865
I'd like to run it against a past tile just to verify that I didn't accidentally switch x & y anywhere or something.  But I don't actually know where to get the full desi_compute_sky invocation for a particular tile; I couldn't find it in the logs?

**Comment by sbailey** [comment]
Thanks for the quick fix in PR #1865.  Closing issue.
